lib/tpm/tpm_simulator_test.go

```diff
@@ -0,0 +1,230 @@
+// This requires openssl libraries to be installed on the machine and findable
+// by the compiler. On macOS:
+// brew install openssl
+// export C_INCLUDE_PATH="$(brew --prefix openssl)/include"
+// export LIBRARY_PATH="$(brew --prefix openssl)/lib"
+// go test ./lib/tpm -run TestWithSimulator -tags tpmsimulator
+
+func TestWithSimulator(t *testing.T) {
+	t.Parallel()
+	ctx := context.Background()
+	log := slog.Default()
+
+	ca := &x509.Certificate{
+		SerialNumber:          big.NewInt(1),
+		BasicConstraintsValid: true,
+		IsCA:                  true,
+		NotBefore:             time.Now().Add(-time.Hour),
+		NotAfter:              time.Now().Add(time.Hour),
+	}
+	caPrivKey, err := rsa.GenerateKey(rand.Reader, 4096)
+	require.NoError(t, err)
+	caBytes, err := x509.CreateCertificate(
+		rand.Reader, ca, ca, &caPrivKey.PublicKey, caPrivKey,
+	)
+	require.NoError(t, err)
+	caPEM := pem.EncodeToMemory(
+		&pem.Block{Type: "CERTIFICATE", Bytes: caBytes},
+	)
+
+	sim, err := tpmsimulator.GetWithFixedSeedInsecure(0)
+	require.NoError(t, err)
+	// This is the EKPubHash that results from the EK generated with the seed 0.
+	wantEKPubHash := "1b5bbe2e96054f7bc34ebe7ba9a4a9eac5611c6879285ceff6094fa556af485c"
+
+	attestTPM, err := attest.OpenTPM(&attest.OpenConfig{
+		TPMVersion:     attest.TPMVersion20,
+		CommandChannel: &fakeCmdChannel{ReadWriteCloser: sim},
+	})
+	require.NoError(t, err)
+	t.Cleanup(func() {
+		assert.NoError(t, attestTPM.Close())
+	})
+
+	t.Run("Without EKCert", func(t *testing.T) {
+		queryRes, attParams, solve, _, err := attestWithTPM(ctx, log, attestTPM)
+		require.NoError(t, err)
+
+		// Check QueryRes looks right.
+		require.Equal(t, wantEKPubHash, queryRes.EKPubHash)
+		require.NotEmpty(t, queryRes.EKPub)
+		require.False(t, queryRes.EKCertPresent)
+		require.Empty(t, queryRes.EKCertSerial)
+		require.Empty(t, queryRes.EKCert)
+
+		// Now attempt a validation that should succeed
+		validated, err := Validate(ctx, log, ValidateParams{
+			EKKey:        queryRes.EKPub,
+			EKCert:       nil,
+			AttestParams: *attParams,
+			Solve:        solve,
+			AllowedCAs:   []string{},
+		})
+		require.NoError(t, err)
+		require.Equal(t, wantEKPubHash, validated.EKPubHash)
+		require.False(t, validated.EKCertVerified)
+		require.Empty(t, validated.EKCertSerial)
+
+		// Attempt a validation which should fail because there's no EKCert and
+		// AllowedCAs is configured.
+		_, err = Validate(ctx, log, ValidateParams{
+			EKKey:        queryRes.EKPub,
+			EKCert:       nil,
+			AttestParams: *attParams,
+			Solve:        solve,
+			AllowedCAs:   []string{string(caPEM)},
+		})
+		require.ErrorContains(t, err, "tpm did not provide an EKCert to validate against allowed CAs")
+	})
+
+	// Write fake EKCert to the TPM
+	origEK, err := attestTPM.EKs()
+	require.NoError(t, err)
+	fakeEKCert := &x509.Certificate{
+		SerialNumber: big.NewInt(1337133713371337),
+		NotBefore:    time.Now().Add(-time.Hour),
+		NotAfter:     time.Now().Add(time.Hour),
+	}
```

Comment:
Suggestion:

```suggestion
	const ekCertSerialNum = 1337133713371337
	const ekCertSerialHex = "04:c0:1d:b4:00:b0:c9"
	fakeEKCert := &x509.Certificate{
		SerialNumber: big.NewInt(ekCertSerialNum),
		NotBefore:    time.Now().Add(-time.Hour),
		NotAfter:     time.Now().Add(time.Hour),
	}
```

and then use the serial constant below?